src/main/java/io/cdap/plugin/batch/source/BSONObjectToRecordTransformer.java

```diff
@@ -0,0 +1,170 @@
+        ? field.getSchema().getNonNullable().getLogicalType()
+        : field.getSchema().getLogicalType();
+
+      if (fieldLogicalType == null) {
+        // Set values of simple types
+        builder.set(field.getName(), value);
+        continue;
+      }
+
+      // Set values of logical types properly
+      switch (fieldLogicalType) {
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          builder.setTimestamp(field.getName(), (ZonedDateTime) value);
+          break;
+        case DECIMAL:
+          builder.setDecimal(field.getName(), (BigDecimal) value);
+          break;
+        default:
+          throw new UnexpectedFormatException("Field type '" + fieldLogicalType + "' is not supported.");
+      }
+    }
+    return builder.build();
+  }
+
+  private Object extractValue(Object object, final Schema schema) {
+    if (object == null || object instanceof BsonUndefined) {
+      return null;
+    }
+
+    Schema.LogicalType fieldLogicalType = schema.isNullable()
+      ? schema.getNonNullable().getLogicalType()
+      : schema.getLogicalType();
+    if (fieldLogicalType != null) {
+      switch (fieldLogicalType) {
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          Date date = (Date) object;
+          return date.toInstant().atZone(ZoneId.ofOffset("UTC", ZoneOffset.UTC));
+        case DECIMAL:
+          Decimal128 decimal = (Decimal128) object;
+          int scale = schema.isNullable() ? schema.getNonNullable().getScale() : schema.getScale();
+          return decimal.bigDecimalValue().setScale(scale, BigDecimal.ROUND_HALF_EVEN);
+        default:
+          throw new UnexpectedFormatException("Field type '" + fieldLogicalType + "' is not supported.");
+      }
+    }
+
+    Schema.Type fieldType = schema.isNullable() ? schema.getNonNullable().getType() : schema.getType();
+    switch (fieldType) {
+      case BOOLEAN:
+        return (Boolean) object;
+      case INT:
+        return (Integer) object;
+      case DOUBLE:
+        return (Double) object;
+      case BYTES:
+        if (object instanceof ObjectId) {
+          // ObjectId can be read as a bytes. Exactly 12 bytes are required
+          ByteBuffer buffer = ByteBuffer.allocate(12);
+          ObjectId objectId = (ObjectId) object;
+          objectId.putToByteBuffer(buffer);
+          return buffer;
+        } else {
+          return (byte[]) object;
+        }
+      case LONG:
+        return (Long) object;
+      case STRING:
+        if (object instanceof ObjectId) {
+          // ObjectId can be read as a string
+          return ((ObjectId) object).toString();
+        } else {
+          return (String) object;
+        }
+      case MAP:
+        Map.Entry<Schema, Schema> mapSchema = schema.isNullable()
+         ? schema.getNonNullable().getMapSchema()
+         : schema.getMapSchema();
+       if (mapSchema.getKey().getType() != Schema.Type.STRING) {
```

Comment:
this should have been validated already right? No need to check it again here.

Reply:
Done.